### ISSUE TITLE
For-Loop changes the mutability of a variable 

### ISSUE THREAD

**Issue by XRustacean** [rust-lang/rust#49265]
Hello Rustaceans !
                         I'm new to Rust , I was experimenting with loops and vectors ,taking account of mutability I wrote this which worked as desired:

 let v = vec![1,2,3,4];
 let mut vec_index:usize = 0;
    for vec_index in 0..4 {
    	println!("Vec Val:{}",v[vec_index] );
    }

What seemed weird to me was when I removed _mut_ keyword , it compiled successfully.

 let  vec_index:usize = 0;
    for vec_index in 0..4 {
    	println!("Vec Val:{}",v[vec_index] );
    }
**I was expecting that compiler will throw an error regrading mutability**
Also, the output was as expected[No Errors]

I have checked the documentation /Book it had no mention about this being a deliberate design decision

**Comment by sfackler** [comment]
A for loop declares a new variable - it shadows the one you declared.

**Comment by XRustacean** [comment]
Why it is designed as such to shadow the initially declared variable?
For e.g. if in some case I do need to my loop count out of the scope of for loop
(when I removed initial variable declaration and resorted to directly declare in for-loop it wasn't accessible outside the scope)
Wouldn't it be much better if compiler prohibits the use of immutable variables in for-loop?
Was their any pros which outweighs the cons here?

**Comment by sfackler** [comment]
> Why it is designed as such to shadow the initially declared variable?

It behaves like for-each loops do in other languages like Java and C++.

> Wouldn't it be much better if compiler prohibits the use of immutable variables in for-loop?

Again, it isn't using an immutable variable.

Iteration by index is rare in the Rust world - that loop would be written as this instead:

```rust
for val in &v {
    println!("Vec Val:{}", v);
}
```

**Comment by XRustacean** [comment]
Ok got you :')
for-loop creates shadows for its purpose.
Some mod should close down this issue.
Thank you.
